content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+> basic concepts via custom-made backends, which can pervert the original design
+> and make it very hard for newcomers to understand the evil TableGen file.”
+
+This means that you'll be tackling the language itself, and the things built
+with it. Which are often more complicated than the language.
+
+It’s like learning C++ and struggling to use [Boost](https://www.boost.org/).
+Someone might say to you “Boost isn’t required, why not remove it and save
+yourself the hassle?”. As someone new to C++, you might not be aware of the
+boundary between the two of them.
+
+Of course this doesn't help you too much if the project you want to contribute
+to uses Boost. You're stuck dealing with both. In LLVM terms, the TableGen
+language and the backends that consume it are a package deal.
+
+I bring this up so that you can draw a distinction between not understanding
+one or the other. Knowing which one is confusing you is a big advantage
+to finding help.
+
+For any given task there's probably one or two "things built with" you need to
+understand and even then not entirely.
+
+Don't think that the TableGen journey (for lack of a less grand term) ends with
+understanding the many many uses of it. If you could get there it would be an
+end of a sort, but you really don't need to and hardly anyone does. Instead
+put your energy into the things that really interest you.
+
+# Compiler Explorer
+
+Of course we have TableGen in Compiler Explorer! Is a language even real if it’s
+not in Compiler Explorer?
+
+(Of course it is, but if you’re favourite language isn’t, Compiler Explorer has
+[excellent documentation](https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingALanguage.md)
+and friendly maintainers)
+
+Compiler Explorer is a whole bunch of different versions of compilers for
+different languages and different architectures that you can access with just a
+browser tab.
+
+It’s an incredible tool for learning, teaching, triaging, optimising and
+[many more](https://www.youtube.com/watch?v=O5sEug_iaf4) things. I won’t go into
+detail about it here, just a few things about TableGen’s inclusion.
+
+The obvious thing is that `llvm-tblgen` doesn’t emit machine code (though a
+hypothetical backend could) so there is no option to compile to binary or
+execute code.
+
+By default you will see the records printed as plain text. A backend can be
+chosen by adding a compiler option manually or as an “Action” from the
+“Overrides” menu.
+
+It’s important to note that TableGen backends have very specific expectations of
+what will be in the source code. Almost as if you had a C++ compiler that
+wouldn’t compile for Arm unless it saw the symbol `arm_is_cool` somewhere in the
+source code.
+
+In the LLVM monorepo all the required classes are set up for you, but in
+Compiler Explorer they are not. So if you want to experiment with an existing
+backend I suggest providing stub implementations of the classes, or copying some
```

Comment:
So, if you would like to experiment with an existing backend, I suggest you provide stub implementations of the classes, or copy some in from the LLVM project